File: python/data_structures/graph.py
import queue1
class Graph:
	def __init__(self):
		self.adj_lists = {}
		self.vertexes = 0

	def __str__(self):
		return str(self.adj_lists)

	def add_vertex(self,vertex):
		if not self.adj_lists.get(vertex,False):
			self.adj_lists[vertex] = []
			self.vertexes +=1

	def add_edge(self,src,dest):
		if dest in self.adj_lists[src]:
			return
		self.adj_lists[src].append(dest)
		self.adj_lists[dest].append(src)

	def remove_edge(self,v1,v2):
		self.adj_lists[v1].remove(v2)
		self.adj_lists[v2].remove(v1)

	def remove_vertex(self,v1):
		x = self.adj_lists[v1]
		while self.adj_lists[v1]:
			v2 = self.adj_lists[v1][0]
			self.remove_edge(v1,v2)
		self.adj_lists.pop(v1)
		self.vertexes -=1

	def dfs(self,start,visited=None):
		stack = [start]
		if visited is None:
			visited = {start: 1}
		result = []
		while stack:
			current = stack.pop()
			result.append(current)
			for neighbor in self.adj_lists[current]:
				if not visited.get(neighbor):
					stack.append(neighbor)
					visited[neighbor] = 1

		return result

	def valid_path(self,start,end):
		result = self.dfs(start)
		if end in result:
			return True
		else:
			return False

	def sections(self):
		components = 0
		visited = {k:False for k in self.adj_lists.keys()}
		for vertex in self.adj_lists.keys():
			if not visited[vertex]:
				components += 1
				self.dfs(vertex,visited)
		return components

	def bfs(self,s,t):
		n = len(self.adj_lists)
		distances = {k:n for k in self.adj_lists.keys()}
		distances[s] = 0
		q = queue1.Queue()
		q.enqueue(s)
		while len(q):
			v = q.dequeue()
			for u in self.adj_lists[v]:
				if distances[u] == n:
					q.enqueue(u)
					distances[u] = distances[v] + 1

		return -1 if distances[t] == n else distances[t]

	def is_bipartite(self,start=None):
		#bugged
		visited = {}
		partition = {}
		if start is None:
			start = list(self.adj_lists.keys())[0]
		for k in self.adj_lists.keys():
			visited[k] = False
			partition[k] = -1

		visited[start] = True
		partition[start] = 0
		q = queue1.Queue()
		q.enqueue(start)
		while len(q):
			v = q.dequeue()
			for u in self.adj_lists[v]:
				if partition[u] == partition[v]:
					return 0
				else:
					if not visited[u]:
						visited[u] = True
						partition[u] = 1 - partition[v]
						q.enqueue(u)
		return 1

if __name__ == '__main__':
	g = Graph()
	g.add_vertex('a')
	g.add_vertex('b')
	g.add_vertex('c')
	g.add_vertex('d')
	g.add_vertex('e')
	g.add_vertex('f')
	g.add_edge('a','b')
	g.add_edge('a','c')
	g.add_edge('b','d')
	g.add_edge('c','e')
	g.add_edge('d','e')
	g.add_edge('d','f')
	g.add_edge('e','f')
	print(g)
	g.remove_vertex('b')
	print(g)
	print(g.dfs('a'))
	g = Graph()
	g.add_vertex('a')
	g.add_vertex('b')
	g.add_vertex('c')
	g.add_vertex('d')
	g.add_edge('a','b')
	g.add_edge('c','b')
	print('vp',g.valid_path('a','d'))
	print('sections',g.sections())
	g.add_edge('a','d')
	print('vertices_passed',g.bfs('b','d'))
	print('bipartite',g.is_bipartite('a'))

File: js/data_structures/graph.js
class Graph{
	constructor(){
		this._adj_lists = {};
		this._vertexes = 0;
	}
	addVertex(vertex){
		if(! this._adj_lists[vertex]){
			this._adj_lists[vertex] = []
			this._vertexes++;
		}
	}
	addEdge(src, dest){
		if(dest in this._adj_lists[src]){
			return;
		}
		this._adj_lists[src].push(dest);
		this._adj_lists[dest].push(src);
	}

	removeEdge(v1, v2){
		this._adj_lists[v1] = this._adj_lists[v1].filter(x => x !== v2);
		this._adj_lists[v2] = this._adj_lists[v2].filter(x => x !== v1);
	}

	removeVertex(v1){
		while(this._adj_lists[v1].length !== 0){
			let v2 = this._adj_lists[v1][0];
			this.removeEdge(v1,v2);
		}
		delete this._adj_lists[v1];
	}

	dfs(start){
		let stack = [start];
		let visited = {start:1};
		let result = [];
		while(stack.length !== 0){
			let current = stack.pop();
			result.push(current);
			for(let i=0; i<this._adj_lists[current].length; i++){
				let neighbor = this._adj_lists[current][i];
				if(! visited[neighbor]){
					stack.push(neighbor);
					visited[neighbor] = 1;
				}
			}
		}
		return result;
	}
}

if (typeof module !== 'undefined' && require.main === module) {
	let g = new Graph();
	g.addVertex('a');
	g.addVertex('b');
	g.addVertex('c');
	g.addVertex('d');
	g.addVertex('e');
	g.addVertex('f');
	g.addEdge('a','b');
	g.addEdge('a','c');
	g.addEdge('b','d');
	g.addEdge('c','e');
	g.addEdge('d','e');
	g.addEdge('d','f');
	g.addEdge('e','f');
	console.log(g);
	g.removeVertex('b');
	console.log(g.dfs('a'));
}
